File: CRECSharpInterpreter/Tests/Good/Loops.cs
using static CRECSharpInterpreter.VarType;

namespace CRECSharpInterpreter.Tests.Good
{
    public class Loops : ITest
    {
        public Loops(string pathNoExt)
        {
            PathNoExt = pathNoExt;
        }

        public string PathNoExt { get; init; }

        public Variable[][] Stack => 
            new[]
            {
                new Variable[]
                {
                    new(@bool, "whileTest", false, true),
                    new(@int, "whileTest2", 0, true),
                    new(@int.Array, "copyArray", 1, true),
                    new(@int.Array, "pasteArray", 5, true),
                    new(@int, "index", 3, true),
                    new(@char.Array, "helloChars", 9, true),
                    new(@char.Array, "helloCharsCopy", 15, true),
                    new(@bool, "arraysAreEqual", true, true),
                    new(@char.Array, "helloBackwards", 21, true),
                    new(@bool, "breakTest", true, true),
                    new(@int, "continueTest", 0, true),
                    new(@int, "toInitialise", 5, true),
                    new(@int, "anInt", 9, true)
                }
            };

        public Variable?[] Heap =>
            new Variable?[]
            {
                new(null),
                new(@int, 3),
                new(@int, 1),
                new(@int, 2),
                new(@int, 3),
                new(@int, 3),
                new(@int, 1),
                new(@int, 2),
                new(@int, 3),
                new(@int, 5),
                new(@char, 'h'),
                new(@char, 'e'),
                new(@char, 'l'),
                new(@char, 'l'),
                new(@char, 'o'),
                new(@int, 5),
                new(@char, 'h'),
                new(@char, 'e'),
                new(@char, 'l'),
                new(@char, 'l'),
                new(@char, 'o'),
                new(@int, 5),
                new(@char, 'o'),
                new(@char, 'l'),
                new(@char, 'l'),
                new(@char, 'e'),
                new(@char, 'h'),
            };

        public Error Error { get; } = Error.None;
    }
}

File: Files/Tests/Loops/Loops.java
boolean whileTest = true ;
while ( whileTest )
    whileTest = false ;

int whileTest2 = 4 ;
while ( whileTest2 > 0 )
    whileTest2 = whileTest2 - 1 ;

int [ ] copyArray = new int [ ] { 1 , 2 , 3 } ;
int [ ] pasteArray = new int [ copyArray . length ] ;
int index = 0 ;
while ( index < copyArray . length )
{
    pasteArray [ index ] = copyArray [ index ] ;
    index = index + 1 ;
}

char [ ] helloChars = new char [ ] { 'h' , 'e' , 'l' , 'l' , 'o' } ;
char [ ] helloCharsCopy = new char [ helloChars . length ] ;
for ( int i = 0 ; i < helloChars . length ; i = i + 1 )
    helloCharsCopy [ i ] = helloChars [ i ] ;

boolean arraysAreEqual = true ;
while ( true )
{
    for ( int i = 0 ; i < 2 ; i = i + 1 )
        continue ;
    if ( helloChars . length != helloCharsCopy . length )
    {
        arraysAreEqual = false ;
        break ;
    }
    for ( int i = 0 ; i < helloChars . length ; i = i + 1 )
    {
        if ( helloChars [ i ] != helloCharsCopy [ i ] )
        {
            arraysAreEqual = false ;
            break ;
        }
    }
    break ;
}

char [ ] helloBackwards = new char [ helloChars . length ] ;
for ( int i = 0 ; i < helloChars . length ; i = i + 1 )
    helloBackwards [ i ] = helloChars [ helloChars . length - i - 1 ] ;

for ( int i = 0 ; ; i = i + 1 )
{
    if ( helloChars [ i ] == 'o' )
        break ;
    if ( helloChars [ i ] == 'l' )
        continue ;
    while ( true )
        break ;
}

boolean breakTest = true ;
while ( breakTest )
    break ;

int continueTest = 3 ;
while ( continueTest > 0 )
{
    int decrement = 1 ;
    continueTest = continueTest - decrement ;
    continue ;
    continueTest = continueTest + decrement ;
}

int toInitialise ;
int anInt = 5 ;
while ( true )
{
    int someInt = anInt ;
    if ( someInt > 7 )
    {
        int [ ] ints = new int [ ] { someInt , 1 } ;
        int second = ints [ 1 ] ;
        if ( second >= 1 )
        {
            if ( someInt == 8 )
            {
                anInt = anInt + 1 ;
                continue ;
            }
            toInitialise = 5 ;
            break ;
        }
    }
    someInt = someInt + 1 ;
    anInt = someInt ;
}